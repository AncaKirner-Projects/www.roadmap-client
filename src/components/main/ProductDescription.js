import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { withStyles } from '@material-ui/core/styles';
import { connect } from 'react-redux';
import { compose } from 'recompose';
import Grid from '@material-ui/core/Grid';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import styles from '../layers/styles/productDescriptionStyles';
import AlertDialog from './AlertDialog';
import ActionTypes from '../../store/actions/actionTypes';
import { addToCart, deleteCartStatusMessage } from '../../store/actions/cartActions';

class ProductDescription extends Component {
  state = {
    openAlertDialog: false
  }

  handleAlertDialogClose = () => {
    this.props.deleteCartStatusMessage(ActionTypes.ADD_TO_CART);
    this.setState({ openAlertDialog: false });
  }

  handleAddToCart = (product) => {
    this.props.addToCart(product);
    this.setState({ openAlertDialog: true });
  }

  render() {
    const { classes, products, status } = this.props;
    const { id } = this.props.match.params;
    const product = products && products.find(elem => elem.id === parseInt(id, 0));
    const addProductStatus = status && status.find(elem => elem.type === ActionTypes.ADD_TO_CART);

    if (!product) return null;

    return (
      <Paper className={classes.root}>
        <Grid container spacing={16}>
          <Grid item xs={12} sm container>
            <Grid item xs container direction="column" spacing={16}>
              <Grid item xs>
                <Typography gutterBottom variant="h5">
                  {product.prod_name}
                </Typography>
                <Typography gutterBottom variant="subtitle1">{product.prod_description}</Typography>
                <Typography color="textSecondary">
                  ID:
                  {product.id}
                </Typography>
              </Grid>
            </Grid>
            <Grid item>
              <Typography variant="subtitle1" align="center">
                {product.price}
                RON
              </Typography>
              <Button variant="contained"
                size="small"
                color="primary"
                onClick={() => this.handleAddToCart(product)}
              >
                Add to cart
              </Button>
              <AlertDialog
                open={this.state.openAlertDialog}
                handleClose={this.handleAlertDialogClose}
                title="Request Status"
                text={addProductStatus && addProductStatus.message}
              />
            </Grid>
          </Grid>
        </Grid>
      </Paper>
    );
  }
}

const mapStateToProps = state => ({
  products: state.products.all,
  status: state.cart ? state.cart.status : []
});

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators({
    addToCart,
    deleteCartStatusMessage
  }, dispatch)
});

export default compose(
  withStyles(styles),
  connect(mapStateToProps, mapDispatchToProps)
)(ProductDescription);
