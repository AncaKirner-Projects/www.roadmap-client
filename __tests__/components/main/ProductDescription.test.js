import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import Button from '@material-ui/core/Button';
import AlertDialog from '../../../src/components/main/AlertDialog';
import ProductDescription from '../../../src/components/main/ProductDescription';
import ActionTypes from '../../../src/store/actions/actionTypes';

const mockStore = configureMockStore();
const UnwrappedComponent = ProductDescription.WrappedComponent;

describe('<ProductDescription />', () => {
  const store = mockStore({
    classes: {},
    products: {
      all: [
        { id: 1, prod_name: 'Prod1' }
      ]
    },
    cart: {
      status: [
        {
          type: ActionTypes.ADD_TO_CART,
          message: 'Test message'
        }
      ]
    }
  });

  const props = {
    openAlertDialog: false,
    match: {
      params: {
        id: 1
      }
    },
    addToCart: jest.fn(),
    deleteCartStatusMessage: jest.fn()
  };

  it('renders an element', () => {
    const wrapper = shallow(<ProductDescription store={store} {...props} />);
    const component = wrapper.dive().dive();

    expect(toJson(component)).toMatchSnapshot();
  });

  // it('handleAlertDialogClose ', () => {
  //   const wrapper = shallow(<UnwrappedComponent {...props} />);
  // wrapper.find(AlertDialog).simulate('handleClose');

  // expect(props.deleteCartStatusMessage).toHaveBeenCalled();
  // });

  // it('handleAddToCart ', () => {
  //   const wrapper = shallow(<ProductDescription store={store} {...props} />);
  //   const component = wrapper.dive().dive();

  //   component.find(Button).simulate('click', { id: 1, name: 'Prod1' });

  //   expect(props.addToCart).toHaveBeenCalled();
  // });
});
