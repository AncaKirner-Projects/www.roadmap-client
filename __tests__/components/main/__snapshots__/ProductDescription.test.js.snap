// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDescription /> renders an element 1`] = `
<WithStyles(Paper)
  className="Connect-ProductDescription--root-1"
>
  <WithStyles(Grid)
    container={true}
    spacing={16}
  >
    <WithStyles(Grid)
      container={true}
      item={true}
      sm={true}
      xs={12}
    >
      <WithStyles(Grid)
        container={true}
        direction="column"
        item={true}
        spacing={16}
        xs={true}
      >
        <WithStyles(Grid)
          item={true}
          xs={true}
        >
          <WithStyles(Typography)
            gutterBottom={true}
            variant="h5"
          >
            Prod1
          </WithStyles(Typography)>
          <WithStyles(Typography)
            gutterBottom={true}
            variant="subtitle1"
          />
          <WithStyles(Typography)
            color="textSecondary"
          >
            ID:
            1
          </WithStyles(Typography)>
        </WithStyles(Grid)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
      >
        <WithStyles(Typography)
          align="center"
          variant="subtitle1"
        >
          RON
        </WithStyles(Typography)>
        <WithStyles(Button)
          color="primary"
          onClick={[Function]}
          size="small"
          variant="contained"
        >
          Add to cart
        </WithStyles(Button)>
        <AlertDialog
          onHandleClose={[Function]}
          open={false}
          text="Test message"
          title="Request Status"
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</WithStyles(Paper)>
`;
